File: Offline/OfflinePlayer/objc/OfflinePlayer/VideosViewController.m
//
//  VideosViewController.m
//  OfflinePlayer
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

#import <BrightcovePlayerSDK/BrightcovePlayerSDK.h>

#import "BCOVVideo+OfflinePlayer.h"
#import "DownloadManager.h"
#import "Notifications.h"
#import "SettingsViewController.h"
#import "UIAlertController+OfflinePlayer.h"
#import "UITabBarController+OfflinePlayer.h"
#import "VideoManager.h"
#import "VideoTableViewCell.h"

#import "VideosViewController.h"


@interface VideosViewController () <BCOVPlaybackControllerDelegate, BCOVPUIPlayerViewDelegate, VideoTableViewCellDelegate, UITableViewDataSource, UITableViewDelegate>

// View that holds the PlayerUI content
// where the video and controls are displayed
@property (nonatomic, weak) IBOutlet UIView *videoContainerView;

// Table view displaying available videos
// from playlist, and its refresh control
@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic, weak) IBOutlet UIView *headerTableView;
@property (nonatomic, weak) IBOutlet UIView *footerTableView;

@property (nonatomic, strong) UIRefreshControl *refreshControl;

@property (nonatomic, strong) UILabel *headerLabel;
@property (nonatomic, strong) UILabel *footerLabel;

// Brightcove-related objects
@property (nonatomic, strong) BCOVPlaybackService *playbackService;
@property (nonatomic, strong) BCOVPUIPlayerView *playerView;
@property (nonatomic, strong) id<BCOVPlaybackController> playbackController;

@property (nonatomic, assign) BOOL statusBarHidden;

@end


@implementation VideosViewController

- (BOOL)prefersStatusBarHidden
{
    return self.statusBarHidden;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    NSDictionary *options = @{
        kBCOVOfflineVideoManagerAllowsCellularDownloadKey: @(NO),
        kBCOVOfflineVideoManagerAllowsCellularPlaybackKey: @(NO),
        kBCOVOfflineVideoManagerAllowsCellularAnalyticsKey: @(NO)
    };

    [BCOVOfflineVideoManager initializeOfflineVideoManagerWithDelegate:DownloadManager.shared
                                                               options:options];

    self.refreshControl = ({
        UIRefreshControl *refreshControl = [UIRefreshControl new];
        [refreshControl addTarget:self
                           action:@selector(requestContentFromPlaybackService)
                 forControlEvents:UIControlEventValueChanged];
        refreshControl;
    });

    self.tableView.refreshControl = self.refreshControl;
    self.tableView.contentInset = UIEdgeInsetsMake(0, 0, 0, 0);

    self.headerLabel = ({
        CGSize size = self.headerTableView.frame.size;
        CGRect frame = CGRectMake(20, 0, size.width - 40, size.height);
        UILabel *headerLabel = [[UILabel alloc] initWithFrame:frame];
        headerLabel.numberOfLines = 1;
        headerLabel.textAlignment = NSTextAlignmentJustified;
        headerLabel.font = [UIFont systemFontOfSize:16];
        headerLabel.textColor = UIColor.systemGrayColor;
        headerLabel.backgroundColor = UIColor.clearColor;
        headerLabel;
    });

    self.headerTableView.layer.borderColor = [UIColor colorWithWhite:0.9f alpha:1.0].CGColor;
    self.headerTableView.layer.borderWidth = 0.3f;
    [self.headerTableView addSubview:self.headerLabel];

    self.footerLabel = ({
        CGSize size = self.footerTableView.frame.size;
        CGRect frame = CGRectMake(20, 0, size.width - 40, size.height);
        UILabel *footerLabel = [[UILabel alloc] initWithFrame:frame];
        footerLabel.numberOfLines = 1;
        footerLabel.textAlignment = NSTextAlignmentJustified;
        footerLabel.font = [UIFont systemFontOfSize:14];
        footerLabel.textColor = UIColor.systemGrayColor;
        footerLabel.backgroundColor = UIColor.clearColor;
        footerLabel;
    });

    self.footerTableView.layer.borderColor = [UIColor colorWithWhite:0.9f alpha:1.0].CGColor;
    self.footerTableView.layer.borderWidth = 0.3f;
    [self.footerTableView addSubview:self.footerLabel];

    self.playerView = ({
        BCOVPUIPlayerViewOptions *options = [BCOVPUIPlayerViewOptions new];
        options.presentingViewController = self;
        options.automaticControlTypeSelection = YES;

        BCOVPUIPlayerView *playerView = [[BCOVPUIPlayerView alloc]
                                         initWithPlaybackController:nil
                                         options:options
                                         controlsView:nil];

        playerView.delegate = self;

        playerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        playerView.frame = self.videoContainerView.bounds;
        [self.videoContainerView addSubview:playerView];

        playerView;
    });

    self.playbackController = ({
        BCOVPlayerSDKManager *sdkManager = BCOVPlayerSDKManager.sharedManager;

        BCOVFPSBrightcoveAuthProxy *authProxy =
        [[BCOVFPSBrightcoveAuthProxy alloc] initWithPublisherId:nil
                                                  applicationId:nil];

        // You can use the same auth proxy for the offline video manager
        // and the call to create the FairPlay session provider.
        BCOVOfflineVideoManager.sharedManager.authProxy = authProxy;

        BCOVBasicSessionProviderOptions *bspOptions = [BCOVBasicSessionProviderOptions new];
        bspOptions.sourceSelectionPolicy = [BCOVBasicSourceSelectionPolicy sourceSelectionHLSWithScheme:kBCOVSourceURLSchemeHTTPS];
        id<BCOVPlaybackSessionProvider> bsp = [sdkManager createBasicSessionProviderWithOptions:bspOptions];

        id<BCOVPlaybackSessionProvider> fps = [sdkManager createFairPlaySessionProviderWithApplicationCertificate:nil
                                                                                               authorizationProxy:authProxy
                                                                                          upstreamSessionProvider:bsp];

        id<BCOVPlaybackController> playbackController = [sdkManager createPlaybackControllerWithSessionProvider:fps
                                                                                                   viewStrategy:nil];

        playbackController.delegate = self;

        playbackController.autoAdvance = YES;
        playbackController.autoPlay = YES;
        playbackController.allowsBackgroundAudioPlayback = YES;

        self.playerView.playbackController = playbackController;

        playbackController;
    });

    [self requestContentFromPlaybackService];

    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(analyticsStorageFullWarningNotificationReceived)
                                               name:kBCOVOfflineVideoManagerAnalyticsStorageFullWarningNotification
                                             object:nil];

    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(updateStatus:)
                                               name:UpdateStatus
                                             object:nil];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    [NSNotificationCenter.defaultCenter postNotificationName:UpdateStatus
                                                      object:nil];
}

- (void)setStatusBarHidden:(BOOL)statusBarHidden
{
    _statusBarHidden = statusBarHidden;
    self.tabBarController.tabBar.hidden = statusBarHidden;
    [self setNeedsStatusBarAppearanceUpdate];
}

- (void)analyticsStorageFullWarningNotificationReceived
{
    [UIAlertController showWithTitle:@"Analytics storage is full"
                             message:@"Encourage the app user to go online"];
}

- (void)updateStatus:(NSNotification *)notification
{
    NSAssert(NSThread.isMainThread, @"Must update UI on main thread");

    if (self.isVisible)
    {
        BCOVVideo *video = notification.object;
        if (video)
        {
            NSUInteger index = [VideoManager.shared.videos indexOfObject:video];
            NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index
                                                        inSection:0];
            [self.tableView reloadRowsAtIndexPaths:@[indexPath]
                                  withRowAnimation:UITableViewRowAnimationNone];
        }
        else
        {
            [self.tableView reloadData];
        }
    }

    [self.tabBarController updateBadge];
}

- (void)requestContentFromPlaybackService
{
    [self.refreshControl beginRefreshing];

    NSDictionary *configuration = @{ kBCOVPlaybackServiceConfigurationKeyAssetReferenceID: kPlaylistRefId };

    NSDictionary *queryParams = @{ @"limit": @(100), @"offset": @(0) };

    [VideoManager.shared retrievePlaylistWithConfiguration:configuration
                                           queryParameters:queryParams
                                                completion:^(BCOVPlaylist *playlist,
                                                             NSDictionary *jsonResponse,
                                                             NSError *error) {

        [self.refreshControl endRefreshing];

        if (playlist)
        {
            SettingsViewController *settingViewController = self.tabBarController.settingsViewController;
            [VideoManager.shared usePlaylist:playlist.videos
                                 withBitrate:settingViewController.bitrate];

            self.headerLabel.text = playlist.properties[kBCOVPlaylistPropertiesKeyName] ?: @"Offline Player";
            self.footerLabel.text = [NSString stringWithFormat:@"%lu %@",
                                           playlist.count,
                                           (playlist.count != 1 ? @"Videos" : @"Video")];
        }
        else
        {
            NSLog(@"No playlist for Id \"%@\" was found.", @"");

            self.headerLabel.text = @"Offline Player";
            self.footerLabel.text = @"0 Videos";
        }
    }];

}


#pragma mark - BCOVPlaybackControllerDelegate

- (void)playbackController:(id<BCOVPlaybackController>)controller
didAdvanceToPlaybackSession:(id<BCOVPlaybackSession>)session
{
    NSLog(@"ViewController - Advanced to new session.");

    // This method is called when ready to play a new video
    NSLog(@"Session source details: %@", session.source);
}

- (void)playbackController:(id<BCOVPlaybackController>)controller
           playbackSession:(id<BCOVPlaybackSession>)session
  didReceiveLifecycleEvent:(BCOVPlaybackSessionLifecycleEvent *)lifecycleEvent
{
    if ([kBCOVPlaybackSessionLifecycleEventFail isEqualToString:lifecycleEvent.eventType])
    {
        NSError *error = lifecycleEvent.properties[@"error"];
        // Report any errors that may have occurred with playback.
        NSLog(@"ViewController - Playback error: %@", error.localizedDescription);
    }
}


#pragma mark - BCOVPUIPlayerViewDelegate

- (void)playerView:(BCOVPUIPlayerView *)playerView
willTransitionToScreenMode:(BCOVPUIScreenMode)screenMode
{
    self.statusBarHidden = screenMode == BCOVPUIScreenModeFull;
}

#pragma mark - VideoTableViewCellDelegate

- (void)performDownloadForVideo:(BCOVVideo *)video
{
    [DownloadManager.shared doDownloadForVideo:video];
}


#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    return VideoManager.shared.videos.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    VideoTableViewCell *videoCell = [tableView dequeueReusableCellWithIdentifier:@"VideoTableViewCell"
                                                                    forIndexPath:indexPath];

    BCOVVideo *video = VideoManager.shared.videos[indexPath.row];
    [videoCell setupWithVideo:video
                  andDelegate:self];

    return videoCell;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}


#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView
didSelectRowAtIndexPath:(nonnull NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath
                             animated:YES];

    BCOVVideo *video = VideoManager.shared.videos[indexPath.row];
    if (!(TARGET_OS_SIMULATOR && video.usesFairPlay))
    {
        [self.playbackController setVideos:@[ video ]];
    }
    else
    {
        [UIAlertController showWithTitle:@"FairPlay Warning"
                                 message:@"FairPlay only works on actual iOS devices.\n\nYou will not be able to view any FairPlay content in the iOS simulator."];
    }
}

- (CGFloat)tableView:(UITableView *)tableView
heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 75;
}

@end

File: Offline/OfflinePlayer/swift/OfflinePlayer/VideosViewController.swift
//
//  VideosViewController.swift
//  OfflinePlayer
//
//  Copyright © 2024 Brightcove, Inc. All rights reserved.
//

import UIKit

import BrightcovePlayerSDK


final class VideosViewController: UIViewController {

    // View that holds the PlayerUI content
    // where the video and controls are displayed
    @IBOutlet fileprivate weak var videoContainerView: UIView!

    // Table view displaying available videos
    // from playlist, and its refresh control
    @IBOutlet fileprivate weak var tableView: UITableView! {
        didSet {
            tableView.refreshControl = refreshControl
            tableView.contentInset = UIEdgeInsets(top: 0,
                                                  left: 0,
                                                  bottom: 0,
                                                  right: 0)
        }
    }

    @IBOutlet fileprivate weak var headerTableView: UIView! {
        didSet {
            headerTableView.layer.borderColor = UIColor.init(white: 0.9,
                                                             alpha: 1.0).cgColor
            headerTableView.layer.borderWidth = 0.3
            headerTableView.addSubview(headerLabel)
        }
    }

    @IBOutlet fileprivate weak var footerTableView: UIView! {
        didSet {
            footerTableView.layer.borderColor = UIColor.init(white: 0.9,
                                                             alpha: 1.0).cgColor
            footerTableView.layer.borderWidth = 0.3
            footerTableView.addSubview(footerLabel)
        }
    }

    fileprivate lazy var headerLabel: UILabel = {
        let headerLabel = UILabel(frame: CGRect(x: 20,
                                                y: 0,
                                                width: headerTableView.frame.size.width - 40,
                                                height: headerTableView.frame.size.height))
        headerLabel.numberOfLines = 1
        headerLabel.textAlignment = .justified
        headerLabel.font = .boldSystemFont(ofSize: 16)
        headerLabel.textColor = .systemGray
        headerLabel.backgroundColor = .clear
        return headerLabel
    }()

    fileprivate lazy var footerLabel: UILabel = {
        let footerLabel = UILabel(frame: CGRect(x: 20,
                                                y: 0,
                                                width: footerTableView.frame.size.width - 40,
                                                height: footerTableView.frame.size.height))
        footerLabel.numberOfLines = 1
        footerLabel.textAlignment = .justified
        footerLabel.font = .boldSystemFont(ofSize: 14)
        footerLabel.textColor = .systemGray
        footerLabel.backgroundColor = .clear
        return footerLabel
    }()

    // Brightcove-related objects
    fileprivate lazy var playerView: BCOVPUIPlayerView? = {
        let options = BCOVPUIPlayerViewOptions()
        options.presentingViewController = self
        options.automaticControlTypeSelection = true

        guard let playerView = BCOVPUIPlayerView(playbackController: nil,
                                                 options: options,
                                                 controlsView: nil) else {
            return nil
        }

        playerView.delegate = self

        playerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        playerView.frame = videoContainerView.bounds
        videoContainerView.addSubview(playerView)

        return playerView
    }()

    fileprivate lazy var playbackController: BCOVPlaybackController? = {
        guard let sdkManager = BCOVPlayerSDKManager.sharedManager(),
              let offlineManager = BCOVOfflineVideoManager.shared(),
              let authProxy = BCOVFPSBrightcoveAuthProxy(publisherId: nil,
                                                         applicationId: nil),
              let sourcePolicy = BCOVBasicSourceSelectionPolicy.sourceSelectionHLS(withScheme: kBCOVSourceURLSchemeHTTPS) else {
            return nil
        }

        // You can use the same auth proxy for the offline video manager
        // and the call to create the FairPlay session provider.
        offlineManager.authProxy = authProxy

        let bspOptions = BCOVBasicSessionProviderOptions()
        bspOptions.sourceSelectionPolicy = sourcePolicy

        guard let bsp = sdkManager.createBasicSessionProvider(with: bspOptions) else {
            return nil
        }

        let fps = sdkManager.createFairPlaySessionProvider(withApplicationCertificate: nil,
                                                           authorizationProxy: authProxy,
                                                           upstreamSessionProvider: bsp)

        guard let playerView,
              let playbackController = sdkManager.createPlaybackController(with: fps,
                                                                           viewStrategy: nil) else {
            return nil
        }

        playbackController.delegate = self
        playbackController.isAutoAdvance = true
        playbackController.isAutoPlay = true
        playbackController.allowsBackgroundAudioPlayback = true

        playerView.playbackController = playbackController

        return playbackController
    }()

    fileprivate lazy var refreshControl: UIRefreshControl = {
        let refreshControl = UIRefreshControl()
        refreshControl.addTarget(self,
                                 action: #selector(requestContentFromPlaybackService),
                                 for: .valueChanged)
        return refreshControl
    }()

    fileprivate lazy var statusBarHidden = false {
        didSet {
            if let tabBarController {
                tabBarController.tabBar.isHidden = statusBarHidden
            }

            setNeedsStatusBarAppearanceUpdate()
        }
    }

    override var prefersStatusBarHidden: Bool {
        return statusBarHidden
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let options = [
            kBCOVOfflineVideoManagerAllowsCellularDownloadKey: false,
            kBCOVOfflineVideoManagerAllowsCellularPlaybackKey: false,
            kBCOVOfflineVideoManagerAllowsCellularAnalyticsKey: false
        ]

        BCOVOfflineVideoManager.initializeOfflineVideoManager(with: DownloadManager.shared,
                                                              options: options)

        requestContentFromPlaybackService()

        NotificationCenter.default.addObserver(self,
                                               selector: #selector(analyticsStorageFullWarningNotificationReceived),
                                               name: NSNotification.Name.bcovOfflineVideoManagerAnalyticsStorageFullWarning,
                                               object: nil)

        NotificationCenter.default.addObserver(self,
                                               selector: #selector(updateStatus(_:)),
                                               name: OfflinePlayerNotifications.UpdateStatus,
                                               object: nil)

    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        NotificationCenter.default.post(name: OfflinePlayerNotifications.UpdateStatus,
                                        object: nil)
    }

    @objc
    fileprivate func analyticsStorageFullWarningNotificationReceived() {
        UIAlertController.showWith(title: "Analytics storage is full",
                                   message: "Encourage the app user to go online")
    }

    @objc
    fileprivate func updateStatus(_ notification: NSNotification) {
        DispatchQueue.main.async { [self] in
            if isVisible {
                if let video = notification.object as? BCOVVideo,
                   let videoId = video.videoId,
                   let videoIndex = VideoManager.shared.videos.firstIndex(where: { $0.videoId == videoId }) {
                    let indexPath = IndexPath(row: videoIndex, section: 0)
                    tableView.reloadRows(at: [indexPath], with: .none)
                } else {
                    tableView.reloadData()
                }
            }

            if let tabBarController {
                tabBarController.updateBadge()
            }
        }
    }

    @objc
    fileprivate func requestContentFromPlaybackService() {
        refreshControl.beginRefreshing()

        let queryParams = ["limit": 100, "offset": 0]

        let configuration = [kBCOVPlaybackServiceConfigurationKeyAssetReferenceID: kPlaylistRefId]
        VideoManager.shared.retrievePlaylist(with: configuration,
                                             queryParameters: queryParams) {
            [self] (playlist: BCOVPlaylist?, json: [AnyHashable: Any]?, error: Error?) in

            refreshControl.endRefreshing()

            if let playlist,
               let videos = playlist.videos as? [BCOVVideo],
               let tabBarController,
               let settingsViewController = tabBarController.settingsViewController {

                let bitrate = settingsViewController.bitrate
                VideoManager.shared.usePlaylist(videos,
                                                with: bitrate)

                headerLabel.text = playlist.properties[kBCOVPlaylistPropertiesKeyName] as? String ?? "Offline Player"
                footerLabel.text = "\(videos.count) \(videos.count != 1 ? "Videos" : "Video")"
            } else {
                print("No playlist for Id \"\(kPlaylistRefId)\" was found.")

                headerLabel.text = "Offline Player"
                footerLabel.text = "0 Videos"
            }
        }
    }
}


// MARK: - BCOVPlaybackControllerDelegate

extension VideosViewController: BCOVPlaybackControllerDelegate {

    func playbackController(_ controller: BCOVPlaybackController!,
                            didAdvanceTo session: BCOVPlaybackSession!) {
        print("ViewController - Advanced to new session.")

        if let source = session.source {
            print("Session source details: \(source)")
        }
    }

    func playbackController(_ controller: BCOVPlaybackController!,
                            playbackSession session: BCOVPlaybackSession!,
                            didReceive lifecycleEvent: BCOVPlaybackSessionLifecycleEvent!) {

        guard let error = lifecycleEvent.properties[kBCOVPlaybackSessionEventKeyError] as? NSError else {
            return
        }

        // Report any errors that may have occurred with playback.
        print("ViewController - Playback error: \(error.localizedDescription)")
    }
}


// MARK: - BCOVPUIPlayerViewDelegate

extension VideosViewController: BCOVPUIPlayerViewDelegate {

    func playerView(_ playerView: BCOVPUIPlayerView!,
                    willTransitionTo screenMode: BCOVPUIScreenMode) {
        statusBarHidden = screenMode == .full
    }
}


// MARK: - VideoTableViewCellDelegate

extension VideosViewController: VideoTableViewCellDelegate {

    func performDownload(forVideo video: BCOVVideo) {
        DownloadManager.shared.doDownload(forVideo: video)
    }
}


// MARK: - UITableViewDataSource

extension VideosViewController: UITableViewDataSource {

    func tableView(_ tableView: UITableView,
                   numberOfRowsInSection section: Int) -> Int {
        return VideoManager.shared.videos.count
    }

    func tableView(_ tableView: UITableView,
                   cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let videoCell = tableView.dequeueReusableCell(withIdentifier: "VideoTableViewCell",
                                                            for: indexPath) as? VideoTableViewCell else {
            return UITableViewCell()
        }

        let video = VideoManager.shared.videos[indexPath.row]
        videoCell.setup(with: video,
                        and: self)

        return videoCell
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
}


// MARK: - UITableViewDelegate

extension VideosViewController: UITableViewDelegate {

    func tableView(_ tableView: UITableView,
                   didSelectRowAt indexPath: IndexPath) {

        tableView.deselectRow(at: indexPath, animated: true)

        let video = VideoManager.shared.videos[indexPath.row]

        if let playbackController {
            if !(UIDevice.current.isSimulator && video.usesFairPlay) {
                playbackController.setVideos([video] as NSFastEnumeration)
            } else {
                UIAlertController.showWith(title: "FairPlay Warning",
                                           message: "FairPlay only works on actual iOS devices.\n\nYou will not be able to view any FairPlay content in the iOS simulator.")
            }
        }
    }

    func tableView(_ tableView: UITableView,
                   heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 75
    }
}
